moving docs to mkdocs (#856)
samuelcolvin/pydantic: docs/usage/models.md

The primary means of defining objects in *pydantic* is via models 
(models are simply classes which inherit from `BaseModel`).

You can think of models as similar to types in strictly typed languages or the requirements of a single endpoint
in an API.

Untrusted data can be passed to a model, after parsing and validation *pydantic* guarantees that the fields
of the resultant model instance will conform to the field types defined on the model.

!!! note
    *pydantic* is primarily a parsing library, **not a validation library**.

>     Validation is a means to and end - building a model which forms to the types and constraints provided.

Reviewer: ```suggestion
    Validation is a means to an end - building a model which conforms to the types and constraints provided.
```

>     In other workds *pydantic* guarentees the types and constraints of the output model not the input data.

Reviewer: ```suggestion
    In other words *pydantic* guarantees the types and constraints of the output model, not the input data.
```

    This might sound like an esoteric distinction, but it is not - you should read about 
    [Data Conversion](#data-conversion) if you're unsure what this means or how it might effect your usage.

> ## Model basic usage

Reviewer: ```suggestion
## Basic model usage
```

```py
from pydantic import BaseModel

class User(BaseModel):
    id: int
    name = 'Jane Doe'
```
`User` here is a model with two fields `id` which is an integer and is required, 

> and `name` which is a string and is not required (it has a a default value). The type of `name` is inferred from the

Reviewer: ```suggestion
and `name` which is a string and is not required (it has a default value). The type of `name` is inferred from the
```

default value, thus a type annotation is not required (however note [this](#field-ordering) warning about field 
order when some fields do not have type annotations).
```py
user = User(id='123')
```
`user` here is an instance of `User`. Initialisation of the object will perform all parsing and validation,
if no `ValidationError` is raised, you know the resulting model instance is valid.
```py
assert user.id == 123
```
fields of a model can be accessed as normal attributes of the user object
the string '123' has been cast to an int as per the field type
```py
assert user.name == 'Jane Doe'
```
name wasn't set when user was initialised, so it has the default value
```py
assert user.__fields_set__ == {'id'}
```
the fields which were supplied when user was initialised:
```py
assert user.dict() == dict(user) == {'id': 123, 'name': 'Jane Doe'}
```
either `.dict()` or `dict(user)` will provide a dict of fields, but `.dict()` can take numerous other arguments.
```py
user.id = 321
assert user.id == 321
```

> This model is mutable so field can be changed.

Reviewer: ```suggestion
This model is mutable so field values can be changed.
```

### Model properties

> The example above only shows the tip of the iceberg of what models can do, 

Reviewer: ```suggestion
The example above only shows the tip of the iceberg of what models can do. 
```

> models contains the following methods and attributes:

Reviewer: ```suggestion
Models contains the following methods and attributes:
```

special key word arguments `__config__` and `__base__` can be used to customise the new model. This includes
extending a base model with extra fields.

```py
{!./examples/dynamic_inheritance.py!}
```

## Custom Root Types

Pydantic models which do not represent a `dict` ("object" in JSON parlance) can have a custom
root type defined via the `__root__` field. The root type can of any type: list, float, int etc.

The root type can be defined via the type hint on the `__root__` field.
The root value can be passed to model `__init__` via the `__root__` keyword argument or as
the first and only argument to `parse_obj`.

```py
{!examples/custom_root_field.py!}
```

## Faux Immutability

Models can be configured to be immutable via `allow_mutation = False` this will prevent changing attributes of
a model. See [model config](model_config.md) for more details on `Config`.

!!! warning
    Immutability in python is never strict. If developers are determined/stupid they can always
    modify a so-called "immutable" object.

```py
{!./examples/mutation.py!}
```

Trying to change `a` caused an error and it remains unchanged, however the dict `b` is mutable and the

> immutability of `foobar` doesn't stop being changed.

Reviewer: ```suggestion
immutability of `foobar` doesn't stop `b` from being changed.
```

## Abstract Base Classes

Pydantic models can be used alongside Python's
[Abstract Base Classes](https://docs.python.org/3/library/abc.html) (ABCs).

```py
{!./examples/ex_abc.py!}
```
_(This script is complete, it should run "as is")_

## Field Ordering

> Field order is important in models since for the following reason:

Reviewer: ```suggestion
Field order is important in models for the following reason:
```
